google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterOperation.java

```diff
@@ -0,0 +1,174 @@
+import java.util.concurrent.Executor;
+
+/**
+ * Represents a single write for BulkWriter, encapsulating operation dispatch and error handling.
+ */
+class BulkWriterOperation {
+  private final SettableApiFuture<WriteResult> future = SettableApiFuture.create();
+  private final DocumentReference documentReference;
+  private final BulkWriter.OperationType operationType;
+  private final ApiFunction<BulkWriterOperation, Void> scheduleWriteCallback;
+  private final Executor successExecutor;
+  private final BulkWriter.WriteResultCallback successListener;
+  private final Executor errorExecutor;
+  private final BulkWriter.WriteErrorCallback errorListener;
+
+  private int failedAttempts;
+
+  /**
+   * @param documentReference The document reference being written to.
+   * @param operationType The type of operation that created this write.
+   * @param scheduleWriteCallback The callback used to schedule a new write.
+   * @param successExecutor The executor to run the user-provided success handler on.
+   * @param successListener The user-provided success handler.
+   * @param errorExecutor The executor to run the user-provided error handler on.
+   * @param errorListener The user-provided error handler.
+   */
+  BulkWriterOperation(
+      DocumentReference documentReference,
+      BulkWriter.OperationType operationType,
+      ApiFunction<BulkWriterOperation, Void> scheduleWriteCallback,
+      Executor successExecutor,
+      BulkWriter.WriteResultCallback successListener,
+      Executor errorExecutor,
+      BulkWriter.WriteErrorCallback errorListener) {
+    this.documentReference = documentReference;
+    this.operationType = operationType;
+    this.scheduleWriteCallback = scheduleWriteCallback;
+    this.successExecutor = successExecutor;
+    this.successListener = successListener;
+    this.errorExecutor = errorExecutor;
+    this.errorListener = errorListener;
+  }
+
+  /**
+   * Returns an ApiFuture that resolves when the operation completes (either successfully or failed
+   * after all retry attempts are exhausted).
+   */
+  public ApiFuture<WriteResult> getFuture() {
+    return future;
+  }
+
+  public DocumentReference getDocumentReference() {
+    return documentReference;
+  }
+
+  /** Callback invoked when an operation attempt fails. */
+  public ApiFuture<Void> onException(FirestoreException exception) {
+    ++failedAttempts;
+
+    final BulkWriterException bulkWriterException =
+        new BulkWriterException(
+            exception.getStatus(),
+            exception.getMessage(),
+            documentReference,
+            operationType,
+            failedAttempts);
+
+    final SettableApiFuture<Void> future = SettableApiFuture.create();
+
+    ApiFutures.addCallback(
+        invokeUserErrorCallback(bulkWriterException),
+        new ApiFutureCallback<Boolean>() {
+          @Override
+          public void onFailure(Throwable throwable) {
+            BulkWriterOperation.this.future.setException(throwable);
```

Comment:
If you rename the instance `future` variable, can you remove the need to use `BulkWriterOperation.this`?